# **Práctica 1: Análisis Exploratorio de Datos (EDA)**

### *Introducción*

##### En el sector bancario, garantizar decisiones acertadas en la evaluación de solicitudes de préstamos es trascendental para minimizar riesgos y maximizar el beneficio. En esta práctica, se examinará un conjunto de datos relacionado con solicitudes de préstamos, empleando técnicas de Análisis Exploratorio de Datos (EDA) vistas en clase. 

### *Objetivos*

##### - Identificar y determinar patrones en los datos, los cuales indiquen la capacidad de los solicitantes para cumplir con sus obligaciones financieras. Ratificando que los usuarios capaces de saldar el préstamo no sean rechazados, a la par de detectar perfiles con dificultades para cubrir la deuda. 

#### - Responder a la pregunta clave: **¿Hay algún tipo de clientes más propenso a no devolver un préstamo?**, guiando al banco en la toma de decisiones para mitigar exposiciones.

### *Desarrollo notebook 1*

### **Importar librerías**

In [2]:
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import plotly.express as px

import sys
sys.path.append('/Users/nicolasantondelavilla/Desktop/p1_aautomatico/practica1')
from funciones import funciones_auxiliares as f_aux

pd.set_option("display.max_rows", 10000)
pd.set_option("display.max_columns", 10000)
pd.set_option("display.width", 10000)

## **Carga y lectura de dataset**

In [3]:
# Situamos a la variable 'SK_ID_CURR' como índice, con el fin de acceder fácilmente a los datos usando estos identificadores como claves.
df = pd.read_csv("/Users/nicolasantondelavilla/Desktop/data_practica_EDA/application_data.csv").set_index("SK_ID_CURR")

##### Es importante señalar que, en este caso, se está empleando una *ruta absoluta* debido a que el tamaño de la base de datos es tan grande que GitHub no puede soportarlo. En situaciones contrarias, donde el tamaño del archivo no es un impedimiento, se debe leer el fichero de datos desde una *ruta relativa* que apunte al repositorio de GitHub.

In [4]:
df.head()

TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE            NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  LIVE_REGION_NOT_WORK_REGION  REG_CITY_NOT_LIVE_CITY  REG_CITY_NOT_WORK_CITY  LIVE_CITY_NOT_WORK_CITY       ORGANIZATION_TYPE  EXT_SOURCE_1  EXT_SOURCE_2  EXT_SOURCE_3  APARTMENTS_AVG  BASEMENTAREA_AVG  YEARS_BEGINEXPLUATATION_AVG  YEARS_BUILD_AVG  COMMONAREA_AVG  ELEVATORS_AVG  ENTRANCES_AVG  FLOORSMAX_AVG  FLOORSMIN_AVG  LANDAREA_AVG  LIVINGAPARTMENTS_AVG  LIVINGAREA_AVG  NONLIVINGAPARTMENTS_AVG  NONLIVINGAREA_AVG  APARTMENTS_MODE  BASEMENTAREA_MODE  YEARS_BEGINEXPLUATATION_MODE  YEARS_BUILD_MODE  COMMONAREA_MODE  ELEVATORS_MODE  ENTRANCES_MODE  FLOORSMAX_MODE  FLOORSMIN_MODE  LANDAREA_MODE  LIVINGAPARTMENTS_MODE  LIVINGAREA_MODE  NONLIVINGAPARTMENTS_MODE  NONLIVINGAREA_MODE  APARTMENTS_MEDI  BASEMENTAREA_MEDI  YEARS_BEGINEXPLUATATION_MEDI  YEARS_BUILD_MEDI  COMMONAREA_MEDI  ELEVATORS_MEDI  ENTRANCES_MEDI  FLOORSMAX_MEDI  FLOORSMIN_MEDI  LANDAREA_MEDI  LIVINGAPARTMENTS_MEDI  LIVINGAREA_MEDI  NONLIVINGAPARTMENTS_MEDI  NONLIVINGAREA_MEDI FONDKAPREMONT_MODE  HOUSETYPE_MODE  TOTALAREA_MODE WALLSMATERIAL_MODE EMERGENCYSTATE_MODE  OBS_30_CNT_SOCIAL_CIRCLE  DEF_30_CNT_SOCIAL_CIRCLE  OBS_60_CNT_SOCIAL_CIRCLE  DEF_60_CNT_SOCIAL_CIRCLE  DAYS_LAST_PHONE_CHANGE  FLAG_DOCUMENT_2  FLAG_DOCUMENT_3  FLAG_DOCUMENT_4  FLAG_DOCUMENT_5  FLAG_DOCUMENT_6  FLAG_DOCUMENT_7  FLAG_DOCUMENT_8  FLAG_DOCUMENT_9  FLAG_DOCUMENT_10  FLAG_DOCUMENT_11  FLAG_DOCUMENT_12  FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  FLAG_DOCUMENT_15  FLAG_DOCUMENT_16  FLAG_DOCUMENT_17  FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR
SK_ID_CURR                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          

## **Análisis general de la tabla**

#### *Dimensión*

In [5]:
print(df.shape, df.drop_duplicates().shape)

(307511, 121) (307511, 121)


##### Con base a la anterior línea de código anterior, es posible determinar que en el DataFrame contiene 121 variables y 307,511 observaciones.

##### *Tipos de datos y su respectivo contenido*

##### Con la finalidad de mantener el código limpio y operar de manera eficiente, a partir de esta sección en adelante se emplearán funciones auxiliares. Un documento donde estan registradas las funciones, facilitando la reutilización de código. A continuación se presenta la función tipos_datos( ) con el fin de determinar el tipo de variable y visualizar su respectivo contenido.

In [6]:
f_aux.tipos_datos(df)

TARGET                         int64        Contenido: 1, 0
NAME_CONTRACT_TYPE             object       Contenido: Cash loans, Revolving loans
CODE_GENDER                    object       Contenido: M, F, XNA
FLAG_OWN_CAR                   object       Contenido: N, Y
FLAG_OWN_REALTY                object       Contenido: Y, N
CNT_CHILDREN                   int64        Contenido: 0, 1, 2, 3, 4, 7, 5, 6, 8, 9, 11, 12, 10, 19, 14
AMT_INCOME_TOTAL               float64      Contenido: Más de 30 valores
AMT_CREDIT                     float64      Contenido: Más de 30 valores
AMT_ANNUITY                    float64      Contenido: Más de 30 valores
AMT_GOODS_PRICE                float64      Contenido: Más de 30 valores
NAME_TYPE_SUITE                object       Contenido: Unaccompanied, Family, Spouse, partner, Children, Other_A, nan, Other_B, Group of people
NAME_INCOME_TYPE               object       Contenido: Working, State servant, Commercial associate, Pensioner, Unemployed, Student,

### La primer variable resultante en esta iteración, es la variable objetivo (TARGET) la cual presenta valores 1 y 0, donde estas simbolizan:
#####   1 - Cliente con dificultades de pago: Pago atrasado de más de X días en al menos una de las primeras Y cuotas del préstamo
#####   0 - Cliente sin dificultades de pago


## **Factores a considerar sobre las variables a emplear en el modelo**

##### Analizando las variables, determinamos la existencia de **variables a futuro** en el conjunto de datos, los cuales pueden llegar a afectar los resultados obtenidos con el modelo, debido a que estas variables en específico son basadas en datos historicos, por lo que en el momento de llamar al modelo, no se encontrarán disponibles.
##### - EXT_SOURCE_1 
##### - EXT_SOURCE_2 
##### - EXT_SOURCE_3
##### - OBS_30_CNT_SOCIAL_CIRCLE
##### - DEF_30_CNT_SOCIAL_CIRCLE
##### - OBS_60_CNT_SOCIAL_CIRCLE
##### - DEF_60_CNT_SOCIAL_CIRCLE
##### Las variables de EXT_SOURCE_1, EXT_SOURCE_2 y EXT_SOURCE_3, simbolizan un puntaje normalizado de una fuente de datos **externa**. Las variables de OBS_30_CNT_SOCIAL_CIRCLE, DEF_30_CNT_SOCIAL_CIRCLE, OBS_60_CNT_SOCIAL_CIRCLE y DEF_60_CNT_SOCIAL_CIRCLE son **historiales** de impago en el circulo cercano del cliente (30 o 60 'Day Past Due').

##### En este caso, no eliminaremos las columnas del DataFrame, ya que consideramos que el cliente te proporciona los datos directamente, sin necesidad de recurrir a fuentes externas  o datos historicos para obtener la información. 

## **Exploración variable objetivo**

In [7]:
# Proporción de valores únicos en la variable objetivo TARGET (%)
df_proporcion = df['TARGET']\
        .value_counts(normalize=True)\
        .mul(100).rename("%").reset_index()

# Conteo de valores únicos en la variable
df_conteo = df['TARGET'].value_counts().reset_index()

# Combinamos los df generamos anteriormente
df_proporcion_conteo = pd.merge(df_proporcion, df_conteo, how='inner')
df_proporcion_conteo

TARGET          %   count
0       0  91.927118  282686
1       1   8.072882   24825

In [8]:
# Graficar el diagrama de barras
fig = px.bar(df_proporcion_conteo, x = "TARGET", y = "%", 
             labels = {'TARGET': 'Target', '%': 'Porcentaje'},
             title = "Distribución de la variable objetivo")

# Mostrar los valores de cada tipo de TARGET
fig.update_traces(text = df_proporcion_conteo['%'].round(2), # Obtener dos decimales en la cifra
                  textposition = 'inside', 
                  texttemplate = '%{text}%',  # Muestra el valor como porcentaje
                  )
# Actualiza las propiedades del eje X, sustituyendo los valores 0 y 1 por los tipos de cliente
fig.update_xaxes(tickmode = 'array', tickvals = [0, 1], 
                 ticktext = ['Tipo 0: Cliente sin dificultades de pago', 'Tipo 1: Cliente con dificultades de pago'])
fig.update_yaxes(title_text="Porcentaje")
import plotly.io as pio
pio.renderers.default = "browser"
fig.show()

##### Con el gráfico anterior, es posible determinar que la probabilidad de obtener una observación aleatoria, de un cliente con dificultades de pago es del 8.07%. 

## **Selección de threshold por filas y columnas para eliminar valores missing**.

#### A continuación se presenta la función nulos_columna( ), la cual otorga la cantidad de nulos por columna y su porcentaje de estos respectivamente.

In [9]:
f_aux.nulos_columna(df)

nulos_columnas  porcentaje_columnas
COMMONAREA_AVG                        214865            69.872297
COMMONAREA_MEDI                       214865            69.872297
COMMONAREA_MODE                       214865            69.872297
NONLIVINGAPARTMENTS_AVG               213514            69.432963
NONLIVINGAPARTMENTS_MEDI              213514            69.432963
NONLIVINGAPARTMENTS_MODE              213514            69.432963
FONDKAPREMONT_MODE                    210295            68.386172
LIVINGAPARTMENTS_MEDI                 210199            68.354953
LIVINGAPARTMENTS_AVG                  210199            68.354953
LIVINGAPARTMENTS_MODE                 210199            68.354953
FLOORSMIN_MODE                        208642            67.848630
FLOORSMIN_AVG                         208642            67.848630
FLOORSMIN_MEDI                        208642            67.848630
YEARS_BUILD_MODE                      204488            66.497784
YEARS_BUILD_MEDI                      204488            66.497784
YEARS_BUILD_AVG                       204488            66.497784
OWN_CAR_AGE                           202929            65.990810
LANDAREA_MEDI                         182590            59.376738
LANDAREA_AVG                          182590            59.376738
LANDAREA_MODE                         182590            59.376738
BASEMENTAREA_MODE                     179943            58.515956
BASEMENTAREA_AVG                      179943            58.515956
BASEMENTAREA_MEDI                     179943            58.515956
EXT_SOURCE_1                          173378            56.381073
NONLIVINGAREA_AVG                     169682            55.179164
NONLIVINGAREA_MODE                    169682            55.179164
NONLIVINGAREA_MEDI                    169682            55.179164
ELEVATORS_MEDI                        163891            53.295980
ELEVATORS_AVG                         163891            53.295980
ELEVATORS_MODE                        163891            53.295980
WALLSMATERIAL_MODE                    156341            50.840783
APARTMENTS_AVG                        156061            50.749729
APARTMENTS_MODE                       156061            50.749729
APARTMENTS_MEDI                       156061            50.749729
ENTRANCES_MEDI                        154828            50.348768
ENTRANCES_MODE                        154828            50.348768
ENTRANCES_AVG                         154828            50.348768
LIVINGAREA_AVG                        154350            50.193326
LIVINGAREA_MODE                       154350            50.193326
LIVINGAREA_MEDI                       154350            50.193326
HOUSETYPE_MODE                        154297            50.176091
FLOORSMAX_MODE                        153020            49.760822
FLOORSMAX_AVG                         153020            49.760822
FLOORSMAX_MEDI                        153020            49.760822
YEARS_BEGINEXPLUATATION_MODE          150007            48.781019
YEARS_BEGINEXPLUATATION_AVG           150007            48.781019
YEARS_BEGINEXPLUATATION_MEDI          150007            48.781019
TOTALAREA_MODE                        148431            48.268517
EMERGENCYSTATE_MODE                   145755            47.398304
OCCUPATION_TYPE                        96391            31.345545
EXT_SOURCE_3                           60965            19.825307
AMT_REQ_CREDIT_BUREAU_WEEK             41519            13.501631
AMT_REQ_CREDIT_BUREAU_MON              41519            13.501631
AMT_REQ_CREDIT_BUREAU_HOUR             41519            13.501631
AMT_REQ_CREDIT_BUREAU_DAY              41519            13.501631
AMT_REQ_CREDIT_BUREAU_YEAR             41519            13.501631
AMT_REQ_CREDIT_BUREAU_QRT              41519            13.501631
NAME_TYPE_SUITE                         1292             0.420148
OBS_60_CNT_SOCIAL_CIRCLE                1021             0.332021
DEF_60_CNT_SOCIAL_CIRCLE                1021             0.332021
OBS_

##### Basandonos en la proporción de valores nulos que contienen las variables o bien el threshold, se opto por mantener las variables, debido a que consideramos necesario profundizar más la relevancia de estas, por medio de ver su desempeño con el modelo.

## **Preprocesamiento inicial de algunas variables**

##### En este caso, se transforma una variable categórica en formato de texto a una representación numérica, con el objetivo de facilitar su procesamiento y hacer su manejo más eficiente en el análisis.

In [10]:
dia = { "MONDAY": 1, "TUESDAY": 2, "WEDNESDAY": 3, "THURSDAY": 4, "FRIDAY": 5, "SATURDAY": 6, "SUNDAY": 7}

df['NWEEKDAY_PROCESS_START'] = df['WEEKDAY_APPR_PROCESS_START'].replace(dia)

/var/folders/9v/wlztnjt95tb_50gdffs9gbhr0000gn/T/ipykernel_1840/367575524.py:3: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



In [11]:
# Agreamos la nueva variable generada al DataFrame y posteriormente rectificamos la creación de esta.
nueva_columna = 'NWEEKDAY_PROCESS_START' in df.columns
nueva_columna

True

In [12]:
# Eliminamos la columna preprocesada y rectificamos la eliminación de la variable del DataFrame.
df.drop("WEEKDAY_APPR_PROCESS_START", axis=1, inplace=True)
columna_existe = 'WEEKDAY_APPR_PROCESS_START' in df.columns
columna_existe

False

#### Bajo la premisa de que existen variables booleanas representadas como cadenas de texto, tales como 'YES', 'NO', 'y', 'n', o cualquier otra variación. Se define la funcion valores_booleanos( ), la cual reemplaza estos valores por 1 en los casos de afirmación y por 0 en los casos de negación.

In [13]:
f_aux.valores_booleanos(df)

Columna 'NAME_CONTRACT_TYPE' contiene valores no booleanos, no se convierte.
Columna 'CODE_GENDER' contiene valores no booleanos, no se convierte.


/Users/nicolasantondelavilla/Desktop/p1_aautomatico/practica1/funciones/funciones_auxiliares.py:98: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`

/Users/nicolasantondelavilla/Desktop/p1_aautomatico/practica1/funciones/funciones_auxiliares.py:98: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



Columna 'NAME_TYPE_SUITE' contiene valores no booleanos, no se convierte.
Columna 'NAME_INCOME_TYPE' contiene valores no booleanos, no se convierte.
Columna 'NAME_EDUCATION_TYPE' contiene valores no booleanos, no se convierte.
Columna 'NAME_FAMILY_STATUS' contiene valores no booleanos, no se convierte.
Columna 'NAME_HOUSING_TYPE' contiene valores no booleanos, no se convierte.
Columna 'OCCUPATION_TYPE' contiene valores no booleanos, no se convierte.
Columna 'ORGANIZATION_TYPE' contiene valores no booleanos, no se convierte.
Columna 'FONDKAPREMONT_MODE' contiene valores no booleanos, no se convierte.
Columna 'HOUSETYPE_MODE' contiene valores no booleanos, no se convierte.
Columna 'WALLSMATERIAL_MODE' contiene valores no booleanos, no se convierte.


TARGET NAME_CONTRACT_TYPE CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  HOUR_APPR_PROCESS_START  REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  LIVE_REGION_NOT_WORK_REGION  REG_CITY_NOT_LIVE_CITY  REG_CITY_NOT_WORK_CITY  LIVE_CITY_NOT_WORK_CITY       ORGANIZATION_TYPE  EXT_SOURCE_1  EXT_SOURCE_2  EXT_SOURCE_3  APARTMENTS_AVG  BASEMENTAREA_AVG  YEARS_BEGINEXPLUATATION_AVG  YEARS_BUILD_AVG  COMMONAREA_AVG  ELEVATORS_AVG  ENTRANCES_AVG  FLOORSMAX_AVG  FLOORSMIN_AVG  LANDAREA_AVG  LIVINGAPARTMENTS_AVG  LIVINGAREA_AVG  NONLIVINGAPARTMENTS_AVG  NONLIVINGAREA_AVG  APARTMENTS_MODE  BASEMENTAREA_MODE  YEARS_BEGINEXPLUATATION_MODE  YEARS_BUILD_MODE  COMMONAREA_MODE  ELEVATORS_MODE  ENTRANCES_MODE  FLOORSMAX_MODE  FLOORSMIN_MODE  LANDAREA_MODE  LIVINGAPARTMENTS_MODE  LIVINGAREA_MODE  NONLIVINGAPARTMENTS_MODE  NONLIVINGAREA_MODE  APARTMENTS_MEDI  BASEMENTAREA_MEDI  YEARS_BEGINEXPLUATATION_MEDI  YEARS_BUILD_MEDI  COMMONAREA_MEDI  ELEVATORS_MEDI  ENTRANCES_MEDI  FLOORSMAX_MEDI  FLOORSMIN_MEDI  LANDAREA_MEDI  LIVINGAPARTMENTS_MEDI  LIVINGAREA_MEDI  NONLIVINGAPARTMENTS_MEDI  NONLIVINGAREA_MEDI FONDKAPREMONT_MODE  HOUSETYPE_MODE  TOTALAREA_MODE WALLSMATERIAL_MODE  EMERGENCYSTATE_MODE  OBS_30_CNT_SOCIAL_CIRCLE  DEF_30_CNT_SOCIAL_CIRCLE  OBS_60_CNT_SOCIAL_CIRCLE  DEF_60_CNT_SOCIAL_CIRCLE  DAYS_LAST_PHONE_CHANGE  FLAG_DOCUMENT_2  FLAG_DOCUMENT_3  FLAG_DOCUMENT_4  FLAG_DOCUMENT_5  FLAG_DOCUMENT_6  FLAG_DOCUMENT_7  FLAG_DOCUMENT_8  FLAG_DOCUMENT_9  FLAG_DOCUMENT_10  FLAG_DOCUMENT_11  FLAG_DOCUMENT_12  FLAG_DOCUMENT_13  FLAG_DOCUMENT_14  FLAG_DOCUMENT_15  FLAG_DOCUMENT_16  FLAG_DOCUMENT_17  FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  NWEEKDAY_PROCESS_START
SK_ID_CURR                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

In [14]:
df['EMERGENCYSTATE_MODE'] = (
    df['EMERGENCYSTATE_MODE']
    .fillna('No')  # Reemplaza nulos con un valor predeterminado (por ejemplo, 'No')
    .replace({'Yes': 1, 'No': 0})  # Reemplaza los valores 'Yes' y 'No'
    .astype(int)  # Convierte a enteros
)

## **Tipos de variables: Categoricas, Continuas y Booleanas**

#### Por medio de la funcion clasificar_variables ( ), definimos los tipos de variables, añadiendo una lista de variables no clasificadas para las variables que no entraron dentro de las demás secciones. 

In [15]:
f_aux.clasificar_variables(df)

Variables Booleanas: 36 ['TARGET', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 'FLAG_PHONE', 'FLAG_EMAIL', 'REG_REGION_NOT_LIVE_REGION', 'REG_REGION_NOT_WORK_REGION', 'LIVE_REGION_NOT_WORK_REGION', 'REG_CITY_NOT_LIVE_CITY', 'REG_CITY_NOT_WORK_CITY', 'LIVE_CITY_NOT_WORK_CITY', 'EMERGENCYSTATE_MODE', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9', 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']
Variables Categóricas: 14 ['NAME_CONTRACT_TYPE', 'CODE_GENDER', 'NAME_TYPE_SUITE', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'OR

(['TARGET',
  'FLAG_OWN_CAR',
  'FLAG_OWN_REALTY',
  'FLAG_MOBIL',
  'FLAG_EMP_PHONE',
  'FLAG_WORK_PHONE',
  'FLAG_CONT_MOBILE',
  'FLAG_PHONE',
  'FLAG_EMAIL',
  'REG_REGION_NOT_LIVE_REGION',
  'REG_REGION_NOT_WORK_REGION',
  'LIVE_REGION_NOT_WORK_REGION',
  'REG_CITY_NOT_LIVE_CITY',
  'REG_CITY_NOT_WORK_CITY',
  'LIVE_CITY_NOT_WORK_CITY',
  'EMERGENCYSTATE_MODE',
  'FLAG_DOCUMENT_2',
  'FLAG_DOCUMENT_3',
  'FLAG_DOCUMENT_4',
  'FLAG_DOCUMENT_5',
  'FLAG_DOCUMENT_6',
  'FLAG_DOCUMENT_7',
  'FLAG_DOCUMENT_8',
  'FLAG_DOCUMENT_9',
  'FLAG_DOCUMENT_10',
  'FLAG_DOCUMENT_11',
  'FLAG_DOCUMENT_12',
  'FLAG_DOCUMENT_13',
  'FLAG_DOCUMENT_14',
  'FLAG_DOCUMENT_15',
  'FLAG_DOCUMENT_16',
  'FLAG_DOCUMENT_17',
  'FLAG_DOCUMENT_18',
  'FLAG_DOCUMENT_19',
  'FLAG_DOCUMENT_20',
  'FLAG_DOCUMENT_21'],
 ['NAME_CONTRACT_TYPE',
  'CODE_GENDER',
  'NAME_TYPE_SUITE',
  'NAME_INCOME_TYPE',
  'NAME_EDUCATION_TYPE',
  'NAME_FAMILY_STATUS',
  'NAME_HOUSING_TYPE',
  'OCCUPATION_TYPE',
  'REGION_RATING_CLIE

In [16]:
lista_var_cat, lista_var_con, lista_var_bool, lista_var_no_clasificadas = f_aux.clasificar_variables(df)

Variables Booleanas: 36 ['TARGET', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 'FLAG_PHONE', 'FLAG_EMAIL', 'REG_REGION_NOT_LIVE_REGION', 'REG_REGION_NOT_WORK_REGION', 'LIVE_REGION_NOT_WORK_REGION', 'REG_CITY_NOT_LIVE_CITY', 'REG_CITY_NOT_WORK_CITY', 'LIVE_CITY_NOT_WORK_CITY', 'EMERGENCYSTATE_MODE', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9', 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']
Variables Categóricas: 14 ['NAME_CONTRACT_TYPE', 'CODE_GENDER', 'NAME_TYPE_SUITE', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'OR

#### Con base a las variables no clasificadas, generamos la funcion nueva_clasificar_variables( ), donde se presenta un formato similar a la anterior función, sin embargo, se cuenta con un bucle que itera sobre las variables no clasificadas, añadiendoles al tipo de variable determinado. 

In [17]:
f_aux.nueva_clasificar_variables(df)

Variables Booleanas: 36 ['TARGET', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE', 'FLAG_PHONE', 'FLAG_EMAIL', 'REG_REGION_NOT_LIVE_REGION', 'REG_REGION_NOT_WORK_REGION', 'LIVE_REGION_NOT_WORK_REGION', 'REG_CITY_NOT_LIVE_CITY', 'REG_CITY_NOT_WORK_CITY', 'LIVE_CITY_NOT_WORK_CITY', 'EMERGENCYSTATE_MODE', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9', 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']
Variables Categóricas: 16 ['NAME_CONTRACT_TYPE', 'CODE_GENDER', 'NAME_TYPE_SUITE', 'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS', 'NAME_HOUSING_TYPE', 'OCCUPATION_TYPE', 'REGION_RATING_CLIENT', 'REGION_RATING_CLIENT_W_CITY', 'OR

(['TARGET',
  'FLAG_OWN_CAR',
  'FLAG_OWN_REALTY',
  'FLAG_MOBIL',
  'FLAG_EMP_PHONE',
  'FLAG_WORK_PHONE',
  'FLAG_CONT_MOBILE',
  'FLAG_PHONE',
  'FLAG_EMAIL',
  'REG_REGION_NOT_LIVE_REGION',
  'REG_REGION_NOT_WORK_REGION',
  'LIVE_REGION_NOT_WORK_REGION',
  'REG_CITY_NOT_LIVE_CITY',
  'REG_CITY_NOT_WORK_CITY',
  'LIVE_CITY_NOT_WORK_CITY',
  'EMERGENCYSTATE_MODE',
  'FLAG_DOCUMENT_2',
  'FLAG_DOCUMENT_3',
  'FLAG_DOCUMENT_4',
  'FLAG_DOCUMENT_5',
  'FLAG_DOCUMENT_6',
  'FLAG_DOCUMENT_7',
  'FLAG_DOCUMENT_8',
  'FLAG_DOCUMENT_9',
  'FLAG_DOCUMENT_10',
  'FLAG_DOCUMENT_11',
  'FLAG_DOCUMENT_12',
  'FLAG_DOCUMENT_13',
  'FLAG_DOCUMENT_14',
  'FLAG_DOCUMENT_15',
  'FLAG_DOCUMENT_16',
  'FLAG_DOCUMENT_17',
  'FLAG_DOCUMENT_18',
  'FLAG_DOCUMENT_19',
  'FLAG_DOCUMENT_20',
  'FLAG_DOCUMENT_21'],
 ['NAME_CONTRACT_TYPE',
  'CODE_GENDER',
  'NAME_TYPE_SUITE',
  'NAME_INCOME_TYPE',
  'NAME_EDUCATION_TYPE',
  'NAME_FAMILY_STATUS',
  'NAME_HOUSING_TYPE',
  'OCCUPATION_TYPE',
  'REGION_RATING_CLIE

## Guardar el nuevo CSV

In [18]:
df.to_csv('/Users/nicolasantondelavilla/Desktop/data_practica_EDA/pd_data_initial_preprocessing.csv')